lib/ansible/modules/packaging/os/redhat_subscription.py

```diff
@@ -540,16 +548,16 @@ def update_subscriptions_by_pool_ids(self, pool_ids):
         for p in consumed_pools:
             existing_pools[p.get_pool_id()] = p.QuantityUsed
 
-        serials_to_remove = [p.Serial for p in consumed_pools if pool_ids.get(p.get_pool_id(), 0) != p.QuantityUsed]
-        serials = self.unsubscribe(serials=serials_to_remove)
-
         missing_pools = {}
         for pool_id, quantity in pool_ids.items():
             if existing_pools.get(pool_id, 0) != quantity:
                 missing_pools[pool_id] = quantity
 
         self.subscribe_by_pool_ids(missing_pools)
 
+        serials_to_remove = [p.Serial for p in consumed_pools if pool_ids.get(p.get_pool_id(), 0) != p.QuantityUsed]
+        serials = self.unsubscribe(serials=serials_to_remove)
+
```

Comment:
Is reversing the order here ( from 'unsub then sub' to 'sub then unsub') going to cause any problems?  Thinking about cases where a system/org may be out or almost out of available subs.

Reply:
It looks like that is possible. Good catch. Firsthand (organizationally multiple times), not fun when that happens.

I had changed the order so that any exception for a "pool ID not found" could be caught and handled before anything gets removed -- which is part of what triggered this PR. My initial approach was to create pre-validate functions (which probably avoided this in hindsight); I switched to try-except logic to avoid unnecessary rework and extra regex searches.

Now that you point out these lines and I'm looking more closely, the logic in the comprehension -- from the existing code -- doesn't make sense to me. It's comparing the pool ID to the quantity??

Reply:
I'm focusing on async/tmpdir work right now. I plan to come back to this.

Reply:
@alikins : Okay, getting back to this... Yeah I believe I misread the logic here. It does attempt to protect what you've currently got subscribed, **IF** it matches the regex passed in as a parameter. What it actually boils down to is one bug and one question of what is reasonably expected behavior.

The bug is that if the module unsubscribes pools not matching the regex and then fails to find anything matching the regex, it reports "changed=true" with no explicit warning or failure and moves on.
The expected behavior questions (subjective) are these:

1. Do we want to fail in the event of an unavailable/nonexistent pool ID? I believe this would be consistent with our other modules' behavior.
2. Do we want to fail BEFORE or AFTER `self.unsubscribe(serials=serials_to_remove)`? IMO, it would be more in the spirit of idempotency if we fail before changes are made during that method.

I'm going to update the PR to reflect this and keep more or less the original order.